elekto/controllers/elections.py

```diff
@@ -94,120 +101,158 @@ def elections_voting_page(eid):
 
     # Redirect to thankyou page if already voted
     if F.g.user.id in [v.user_id for v in e.voters]:
-        return F.render_template('errors/message.html',
-                                 title='You have already voted',
-                                 message='To re-cast your vote, please visit\
-                                 the election page.')
-
-    if F.request.method == 'POST':
+        return F.render_template(
+            "errors/message.html",
+            title="You have already voted",
+            message="To re-cast your vote, please visit\
+                                 the election page.",
+        )
+
+    if F.request.method == "POST":
         # encrypt the voter identity
-        passcode = ''.join(secrets.choice(string.digits) for i in range(6))
-        if len(F.request.form['password']):
-            passcode = F.request.form['password']
-        voter = generate_password_hash(F.g.user.username + '+' + passcode)
+        passcode = "".join(secrets.choice(string.digits) for i in range(6))
+        if len(F.request.form["password"]):
+            passcode = F.request.form["password"]
+
+        passcode = passcode.encode("utf-8")
+        kdf = pwhash.argon2i.kdf
+        salt = utils.random(pwhash.argon2i.SALTBYTES)
+        key = kdf(secret.SecretBox.KEY_SIZE, passcode, salt)
+        box = secret.SecretBox(key)
+
+        voter = Voter(user_id=F.g.user.id, salt=salt)
+
+        enc_ballot_ids = []  # list of encrypted ballot ids
+
+        ballot_voter = str(uuid.uuid4())
 
         for k in F.request.form.keys():
-            if k.split('@')[0] == 'candidate':
-                candidate = k.split('@')[-1]
+            if k.split("@")[0] == "candidate":
+                candidate = k.split("@")[-1]
                 rank = F.request.form[k]
-                ballot = Ballot(rank=rank, candidate=candidate, voter=voter)
+                ballot = Ballot(rank=rank, candidate=candidate, voter=ballot_voter)
+                SESSION.add(ballot)
+                SESSION.commit()
+                enc_ballot_id = box.encrypt(str(ballot.id).encode("utf-8"))
+                enc_ballot_ids.append(b64encode(enc_ballot_id).decode("utf-8"))
                 e.ballots.append(ballot)
 
+        voter.ballot_ids = json.dumps(
+            enc_ballot_ids
+        )  # json serialized encrypted ballot ids
```

Comment:
Aren't we storing a UUID for the voter in the `ballot.voter` of all the votes and storing the encryption of that UUID in `voter.encrypted_ballot` instead of `voter.ballot_ids`?

Reply:
@kalkayan I'll do the required modifications as per our discussion.

Reply:
@kalkayan modifications done as per the review via latest updates